grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,314 @@
+                // NOT_SERVING will be sent to clients by changing a healthiness of a server
+                serverHealth.setHealthy(false);
+                synchronized (watchers) {
+                    watchers.values().forEach(entry -> entry.getKey().onCompleted());
+                    watchers.clear();
+                }
+            }
+        });
+    }
+
+    @VisibleForTesting
+    ServingStatus checkServingStatus(@Nullable String serviceName) {
+        if (!isServerHealthy()) {
+            return ServingStatus.NOT_SERVING;
+        }
+        if (Strings.isNullOrEmpty(serviceName)) {
+            return ServingStatus.SERVING;
+        }
+        final ListenableHealthChecker listenableHealthChecker = grpcServiceHealthCheckers.get(serviceName);
+        if (listenableHealthChecker == null) {
+            return ServingStatus.SERVICE_UNKNOWN;
+        }
+        if (listenableHealthChecker.isHealthy()) {
+            return ServingStatus.SERVING;
+        }
+        return ServingStatus.NOT_SERVING;
+    }
+
+    @VisibleForTesting
+    void changeServerStatus(boolean isHealthy) {
+        serverHealth.setHealthy(isHealthy);
+    }
+
+    private void addServerHealthUpdateListener(List<HealthCheckUpdateListener> updateListeners) {
+        serverHealth.addListener(healthChecker -> {
+            updateListeners.forEach(updateListener -> {
+                try {
+                    updateListener.healthUpdated(healthChecker.isHealthy());
+                } catch (Throwable t) {
+                    logger.warn("Unexpected exception from HealthCheckUpdateListener.healthUpdated():", t);
+                }
+            });
+        });
+    }
+
+    private Consumer<String> provideInternalHealthUpdateListener() {
+        return serviceName -> {
+            synchronized (watchers) {
+                final ServingStatus currentStatus = checkServingStatus(serviceName);
+                final HealthCheckResponse healthCheckResponse = getHealthCheckResponse(currentStatus);
+                if (serviceName.isEmpty()) {
+                    notifyStreamObservers("", healthCheckResponse, currentStatus);
+                    return;
+                }
+                notifyStreamObservers(serviceName, healthCheckResponse, currentStatus);
+            }
+        };
+    }
+
+    private void notifyStreamObservers(
+            String serviceName,
+            HealthCheckResponse healthCheckResponse,
+            ServingStatus currentStatus) {
+        final List<Entry<Entry<StreamObserver<HealthCheckResponse>, ServingStatus>,
+                Entry<StreamObserver<HealthCheckResponse>, ServingStatus>>> updated = new ArrayList<>();
+        for (Entry<StreamObserver<HealthCheckResponse>, ServingStatus> streamObserver
+                : watchers.get(serviceName)) {
+            if (streamObserver.getValue() != currentStatus) {
+                streamObserver.getKey().onNext(healthCheckResponse);
+                updated.add(Maps.immutableEntry(
```

Comment:
How about introducing a class that contains `StreamObserver` and `ServingStatus` and just changing the `ServingStatus` when it's updated?

```java
private static class WatcherEntry {
    private final StreamObserver<HealthCheckResponse> responseObserver;
    private final ServingStatus currentStatus;

    WatcherEntry(StreamObserver<HealthCheckResponse> responseObserver, ServingStatus currentStatus) {
        this.responseObserver = responseObserver;
        this.currentStatus = currentStatus;
    }
}
```
This `watcher` map is only used in the synchronized block so I think it's okay to use mutable data which also makes the code a lot simpler.

Reply:
It seems good